glib/src/subclass/mod.rs

```diff
@@ -197,7 +197,220 @@
+//!     assert!(simple_module_object_type.is_valid());
+//! }
+//! ```
+//!
+//! # Example for registering a `glib::Object` subclass within a plugin
+//!
+//! The following code implements a subclass of `glib::Object` and registers it as
+//! a dynamic type.
+//!
+//! ```rust
+//! use glib::prelude::*;
+//! use glib::subclass::prelude::*;
+//!
+//! pub mod imp {
+//!     use std::ffi::CString;
+//!     use super::*;
+//!     use glib::translate::IntoGlib;
+//!     use glib::translate::FromGlib;
+//!
+//!     // SimplePluginObject is a dynamic type.
+//!     #[derive(Default)]
+//!     pub struct SimplePluginObject;
+//!
+//!     #[glib::dynamic_object_subclass(plugin_type = super::SimpleTypePlugin)]
+//!     impl ObjectSubclass for SimplePluginObject {
+//!         const NAME: &'static str = "SimplePluginObject";
+//!         type Type = super::SimplePluginObject;
+//!     }
+//!
+//!     impl ObjectImpl for SimplePluginObject {}
+//!
+//!     // SimpleTypePlugin is the type plugin within the dynamic type is registered.
+//!     #[derive(Default)]
+//!     pub struct SimpleTypePlugin {
+//!         type_info: std::cell::Cell<Option<glib::TypeInfo>>
+//!     }
+//!
+//!     #[glib::object_subclass]
+//!     impl ObjectSubclass for SimpleTypePlugin {
+//!         const NAME: &'static str = "SimpleTypePlugin";
+//!         type Type = super::SimpleTypePlugin;
+//!         type Interfaces = (glib::TypePlugin,);
+//!     }
+//!
+//!     impl ObjectImpl for SimpleTypePlugin {}
+//!
+//!     impl TypePluginImpl for SimpleTypePlugin {
+//!         /// Uses the plugin and registers the dynamic type
+//!         fn use_plugin(&self) {
+//!             SimplePluginObject::on_implementation_load(self.obj().as_ref());
+//!         }
+//!
+//!         /// Unuses the plugin.
+//!         fn unuse_plugin(&self) {
+//!             SimplePluginObject::on_implementation_unload(self.obj().as_ref());
+//!         }
+//!
+//!         fn complete_type_info(&self, _type_: glib::Type) -> (glib::TypeInfo, glib::TypeValueTable) {
+//!             assert!(self.type_info.get().is_some());
+//!             // return type info
+//!             (self.type_info.get().unwrap(), glib::TypeValueTable::default())
+//!         }
+//!     }
+//!
+//!     impl TypePluginRegisterImpl for SimpleTypePlugin {
+//!         fn add_dynamic_interface(&self, _: glib::Type, _: glib::Type, _: &glib::InterfaceInfo) {
+//!             unimplemented!()
+//!         }
+//!
+//!         fn register_dynamic_enum(&self,_: &str, _: &glib::EnumValue) -> glib::Type {
+//!             unimplemented!()
+//!         }
+//!
+//!         fn register_dynamic_flags(&self,_: &str, _: &glib::FlagsValue) -> glib::Type {
+//!             unimplemented!()
+//!         }
+//!
+//!         fn register_dynamic_type(&self, parent_type: glib::Type, type_name: &str, type_info: &glib::TypeInfo, flags: glib::TypeFlags) -> glib::Type {
+//!             let type_ = unsafe {
```

Comment:
I wonder if there wouldn't be some more convenient (even if `unsafe`) API that wouldn't require user code to do FFI directly.

`g_type_from_name()` for example has safe wrappers, `g_type_register_dynamic()` should be possible to wrap in something a bit more convenient at least (taking non-FFI types, taking care of the C string conversion, ...)